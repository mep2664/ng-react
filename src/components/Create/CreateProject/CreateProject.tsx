import * as React from "react";
import { Button, Loader, SelectInput, TextArea, TextInput } from "../..";
import styled from "styled-components";
import gql from "graphql-tag";
import { useQuery, useMutation } from "@apollo/react-hooks";

export const GET_TEAMS = gql`
    query {
        allTeams {
            teamId
            teamName
        }
    }
`;

export const CREATE_PROJECT = gql`
    mutation createProject($projectName: String!, $teamId: ID, $description: String) {
            createProject(projectName: $projectName, teamId: $teamId, description: $description)
            {
                project {
                    projectId
                    projectName
                    description
                }
            }
    }
`;

const ProjectInfo = styled.div`
    display: grid;
    grid-gap: 25px;
    align-items: center;
    justify-items: end;
    grid-template-columns: minmax(auto, 300px) auto;
    box-sizing: border-box;
    padding: 25px;
`;

const DescriptionWrapper = styled.div`
    display: flex;
    flex-direction: column;
    height: 100%;
    width: 100%;
`;

export const CreateProject: React.FC = () => {
    const [projectName, setProject] = React.useState<string>("");
    const [teamId, setTeamId] = React.useState<string>("");
    const [description, setDescription] = React.useState<string>("");
    const { data, error, loading } = useQuery(GET_TEAMS);
    const [createProject] = useMutation(CREATE_PROJECT);

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        const projectData = {
            projectName,
            description,
            teamId,
        }
        createProject({ variables: projectData });

    }

    if (loading) {
        return <Loader />;
    }

    if (error) {
        if (error.networkError && "statusCode" in error.networkError
            && error.networkError["statusCode"] === 405) {
            window.location.assign(`${window.location.protocol}//${window.location.host}/`);
        }
        return <div>{error.message}</div>;
    }

    const teamOptions = data.allTeams.map((team: any) => { return { caption: team.teamName, value: team.teamId } });

    return (
        <div>
            {`Create Project...`}
            <ProjectInfo>
                <form action="localhost:5556/project" method="post" onSubmit={handleSubmit}>
                    <TextInput name="projectName" label="Project Name" value={projectName} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setProject(e.target.value)} />
                    <SelectInput name="teamId" label="Team" options={teamOptions} value={teamId} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setTeamId(e.target.value)} />
                    <DescriptionWrapper>
                        <TextArea name="description" label="Description" value={description} onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => setDescription(e.target.value)} />
                    </DescriptionWrapper>
                    <Button caption="submit" type="submit" emphasis="Primary" />
                </form>
            </ProjectInfo>
        </div>
    );
}