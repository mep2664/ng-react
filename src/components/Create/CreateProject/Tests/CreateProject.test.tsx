import React from "react"
import { render, fireEvent, wait } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect";
import { CreateProject, CREATE_PROJECT, GET_TEAMS } from "../CreateProject"
import { MockedProvider } from '@apollo/react-testing';

jest.mock("../../../Loaders/Loader", () => {
    return {
        Loader: () => <span>loading...</span>
    };
})

const teams = [
    {
        teamId: "5da4f7e39be0c7d43e05950b",
        teamName: "Canyon"
    },
    {
        teamId: "5da4f7e39be0c7d43e05950f",
        teamName: "Ridge"
    },
    {
        teamId: "5da4f7e39be0c7d43e059511",
        teamName: "Peak"
    },
];

const vars = { projectName: "PROJECT_NAME", teamId: "5da4f7e39be0c7d43e05950f", description: "DESCRIPTION" };
let numTimesSubmitted = 0;
const mocks = [
    {
        request: {
            query: GET_TEAMS,
            variables: {},
        },
        result: {
            data: {
                allTeams: teams,
            },
        },
    },
    {
        request: {
            query: CREATE_PROJECT,
            variables: vars,
        },

        result: () => {
            numTimesSubmitted++;
            return {
                data: {
                    createProject: {
                        project: {
                            projectId: "PROJECT_ID",
                            projectName: vars.projectName,
                            description: vars.description,
                        },
                    },
                },
            };
        },
    },
];

describe("Tests for the CreateProject component", () => {
    test("Component matches snapshot", async () => {
        const { container } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateProject />
            </MockedProvider>,
        );
        await wait();
        expect(container).toMatchSnapshot();
    });

    test("Displays loader when waiting for data", () => {
        const { getByText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateProject />
            </MockedProvider>,
        );
        const loader = getByText("loading...");
        expect(loader).toBeDefined();
    });

    test("Displays error when request for data fails", async () => {
        const errorMock = {
            request: {
                query: GET_TEAMS,
                variables: {},
            },
            error: new Error("something went wrong..."),
        };
        const { getByText } = render(
            <MockedProvider mocks={[errorMock]} addTypename={false}>
                <CreateProject />
            </MockedProvider>,
        );
        await wait();
        const error = getByText("Network error: something went wrong...");
        expect(error).toBeDefined();
    });

    test("Changes project name correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateProject />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Project Name");
        fireEvent.change(input, { target: { value: vars.projectName } });
        expect(input).toHaveValue(vars.projectName);
    });

    test("Changes team correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateProject />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Team");
        fireEvent.change(input, { target: { value: vars.teamId } });
        expect(input).toHaveValue(vars.teamId);
    });

    test("Changes description correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateProject />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Description");
        fireEvent.change(input, { target: { value: vars.description } });
        expect(input).toHaveValue(vars.description);
    });

    test("Submits correctly", async () => {
        const { getByText, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateProject />
            </MockedProvider>,
        );
        await wait();

        const projectName = getByLabelText("Project Name");
        fireEvent.change(projectName, { target: { value: vars.projectName } });

        const team = getByLabelText("Team");
        fireEvent.change(team, { target: { value: vars.teamId } });

        const description = getByLabelText("Description");
        fireEvent.change(description, { target: { value: vars.description } });

        const submit = getByText("submit");
        fireEvent.click(submit);
        await wait();
        expect(numTimesSubmitted).toEqual(1);
    });
});
