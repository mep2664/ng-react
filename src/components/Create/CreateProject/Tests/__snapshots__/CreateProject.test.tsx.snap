// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the CreateProject component Component matches snapshot 1`] = `
<div>
  <div>
    Create Project...
    <div
      class="sc-Rmtcm AwuUM"
    >
      <form
        action="localhost:5556/project"
        method="post"
      >
        <label
          class="sc-kEYyzF cXUhpx"
          for="textInput__Project Name"
        >
          Project Name
        </label>
        <input
          class="sc-kkGfuU bvDUFc"
          id="textInput__Project Name"
          name="projectName"
          type="text"
        />
        <label
          class="sc-ckVGcZ envGLI"
          for="selectInput__Team"
        >
          Team
        </label>
        <select
          class="sc-jKJlTe SQmOY"
          id="selectInput__Team"
          name="teamId"
        >
          <option
            value=""
          >
            &lt; Not Selected &gt;
          </option>
          <option
            value="5da4f7e39be0c7d43e05950b"
          >
            Canyon
          </option>
          <option
            value="5da4f7e39be0c7d43e05950f"
          >
            Ridge
          </option>
          <option
            value="5da4f7e39be0c7d43e059511"
          >
            Peak
          </option>
        </select>
        <div
          class="sc-bRBYWo kjpjtk"
        >
          <label
            class="sc-eNQAEJ hALvCQ"
            for="textArea__Description"
          >
            Description
          </label>
          <textarea
            class="sc-hMqMXs elAqlG"
            id="textArea__Description"
            name="description"
          />
        </div>
        <button
          class="sc-bdVaJa eVEMrn"
          type="submit"
        >
          submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
